runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformExecutor.java

```diff
@@ -138,13 +141,6 @@ private void processElements(
 
         evaluator.processElement(value);
 
-        // Report the physical metrics after each element
-        MetricUpdates deltas = metricsContainer.getUpdates();
-        if (deltas != null) {
-          context.getMetrics().updatePhysical(inputBundle, deltas);
```

Comment:
Ideas for other options that preserve functionality and code reduction:

1. Have the DirectRunner report tentative metrics every Nth element (reduces the amount of synchronization).
2. Allow copying the MetricsContainer. Then, the DirectRunner can create a copy and put it on a queue for a separate thread that actually does the update aggregation. This will avoid synchronization.
3. Extending (2) -- I believe it should be safe for multiple threads to *read* a MetricsContainer even if it is being modified by one thread, since it uses Atomic values. Given that, we could just pass a reference to the MetricsContainer to incorporate to the queue, and have it maintain control of the aggregated MetricsContainers.

2 or 3 would also be most like this would happen within a distributed runner, where the actual aggregation was happening outside of the processing thread.